### DISCUSSION TITLE
Unable to fully load a pretrained model when using FastAPI in Docker? Works locally

### DISCUSSION THREAD

**Discussion by uncvrd** [fastapi/fastapi#10640] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

Here's a GitHub repository with a minimum reproducible example: 

https://github.com/uncvrd/clap-mre

Instructions for getting started are in the README (there are 2 steps)

### Description

Hi everyone - I've been trying to containerize a pretrained model so that I can create a web service that returns embeddings generated from this model ([LAION CLAP](https://github.com/LAION-AI/CLAP)). For this task I chose FastAPI. I've been able to get this running perfectly on my local environment, however I've run in to some odd issues when containerizing the web service so that I can deploy it in my infra.

I use FastAPI's `lifespan` method to preload the model when starting up my server so that I don't have to load it on each request. Like this:

```
# Loads the model when the app starts
@asynccontextmanager
async def lifespan(app: FastAPI):
    # anything before yield is loaded on startup
    model = laion_clap.CLAP_Module(enable_fusion=False, amodel="HTSAT-base")
    model.load_ckpt("/clap-data/music_speech_audioset_epoch_15_esc_89.98.pt")  
    yield
    # Anything after yield is run on teardown
```

However, for some reason, when I start FastAPI in the container, only about 75% of the model is loaded before uvicorn says that the application is "ready". Therefore the app fails to generate embeddings when hitting the endpoint. The model i use prints out all the different layers from the pretrained model that have been loaded to the console so that's how I know it's not loading correctly. You can observe this by noticing that it stops at:

`text_branch.encoder.layer.9.output.dense.bias 	 Loaded`

When it really needs to finish at

`audio_projection.2.bias 	 Loaded`

What's really odd is that when I type a change to trigger a reload in my code, you can see the final layers printed to screen before the app exits/restarts. I feel like I'm running in to some sort of memory issue? Like it's flushing out the steps it could't do?

I asked this in the official LAION Discord server (the creators of the pretrained model) and had another user recreate the same problem but he couldn't figure it out either.

<img width="1066" alt="image" src="https://github.com/tiangolo/fastapi/assets/25097099/5c1cdfb2-a1e0-4703-8c64-0f0ce98416f9">

---

What's odd though is that if you simply create a method to load the model without spinning up the webserver within Docker (like I did within `test.py`) it'll load the whole thing no problem so it can't just be a Docker memory issue, right?...

Does anyone have any idea about how I can go about solving this problem? It only happens in Docker, it works great in my local environment which makes me think it's something correlating with uvicorn and how the process manages memory. I'd really love any support. Thanks a lot and let me know what else I can provide to help!

Truly stumped.

### Operating System

macOS

### FastAPI Version

0.104.1

### Pydantic Version

2.4.2

### Python Version

3.11.6

### Additional Context

Here's a video showing this problem: 

https://github.com/tiangolo/fastapi/assets/25097099/3a4288fe-4457-4e3b-a60d-112d203091c8

**Comment by jgould22** [comment]
Sounds like your output is being buffered, the model is loaded, the output is just sitting in a buffer.

Set `PYTHONUNBUFFERED=1` in your docker container and see if it still does it.

https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUNBUFFERED

edit: missed the part where it was not generating embeddings, what happens if you run from something like gunicorn?

**Reply by uncvrd** [reply]
Hey @jgould22 thanks a lot for looking in to this! The unbuffered env variable definitely fixed the printing to screen part. 

Regarding empty response...I was hoping these were correlated but I guess not.

So I've tried doing something like this `gunicorn app.main:app --workers 1 --worker-class uvicorn.workers.UvicornWorker --bind 0.0.0.0:8000` in my Dockerfile and still had the same problem. 

Taken from here: https://fastapi.tiangolo.com/deployment/server-workers/#run-gunicorn-with-uvicorn-workers

Is that sorta what you were asking? Or what would be a better way to try with gunicorn? Not sure if necessary but just a reminder that uvicorn runs this no problem outside of Docker. I've updated my example repo with the ENV variable and suggested FastAPI lifespan pattern if that helps too!

Thanks again for the help so far :)

**Reply by jgould22** [reply]
I am able to run your sample repo and I get an embedding back when I make a request. 

```
[jordan@box ~]$ curl -X 'POST'   'http://0.0.0.0:8002/text-embedding'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d '{
  "queries": [
    "a test string", "another test string"
  ]
}' | jq . 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 21642  100 21577  100    65  91438    275 --:--:-- --:--:-- --:--:-- 92093
{
  "embedding": [
    [
      -0.10073041170835495,
      ...
      ...
      -0.025192435830831528,
      0.0720624029636383,
      0.007903791032731533,
      ...
      0.027038948610424995,
      0.03642858937382698,
      -0.0018659186316654086,
      -0.04501241073012352,
      0.05328139290213585,
      0.0065545919351279736,
      -0.014727740548551083,
      0.08306930214166641
    ]
  ]
}
```

**Reply by uncvrd** [reply]
@jgould22 No way, that is so odd...what environment are you running this on? (Mac Intel/M1/2, Linux)? I just tried the same exact command and the response from curl just hangs

https://github.com/tiangolo/fastapi/assets/25097099/ff12b420-4364-4984-bb6e-20b391178594

**Reply by uncvrd** [reply]
Alright I dusted off my Intel mac and executed the cURL and it ran no problem 😩 This is so strange

On my M2 MacBook:

Local: works
Docker: **FAILS**

On my Intel MacBook:

Local: works
Docker: works

Not even sure where to start debugging this now, but must not be a FastAPI problem, yea?

**Reply by jgould22** [reply]
Docker on the newer arm chips can be a bit finicky.

You will need to make sure all of your stack/packages plays nicely with arm processors since the python container that will get built will be an ARM one.

**Comment by minji2744** [comment]
Hello, I want to ask whether you found a solution to this problem? I'm having a similar issue with loading model in docker image! Thanks!

**Reply by uncvrd** [reply, marked as answer]
> Hello, I want to ask whether you found a solution to this problem? I'm having a similar issue with loading model in docker image! Thanks!

Yes it was because I was building the image on Apple Silicon. If I forced the build for intel architecture it works great
